python/paddle/v2/framework/tests/op_test.py

```diff
@@ -5,175 +5,313 @@
-        var.get_tensor()
-    for output in backward_op.output_vars():
-        var = scope.new_var(output)
-        var.get_tensor()
-    return backward_op
-
-
-def get_gradient(scope, op, inputs, outputs, grad_name, place,
-                 no_grad_set=None):
-    ctx = core.DeviceContext.create(place)
-
-    set_input(scope, op, inputs, place)
-
-    op.infer_shape(scope)
-    op.run(scope, ctx)
-
-    if no_grad_set is None:
-        no_grad_set = set()
-
-    backward_op = get_backward_op(scope, op, no_grad_set)
-    set_output_grad(scope, op, outputs, place)
+    @classmethod
+    def get_numeric_gradient(cls,
+                             scope,
+                             op,
+                             inputs,
+                             input_to_check,
+                             output_names,
+                             delta=0.005,
+                             in_place=False,
+                             strict=False):
+        # compute numeric gradients on CPU
+        cpu_place = core.CPUPlace()
+        cls.feed_input(scope, op, inputs, cpu_place)
+        op.infer_shape(scope)
+        ctx = core.DeviceContext.create(cpu_place)
+
+        tensor_to_check = cls.get_tensor_by_name(scope, input_to_check)
+        tensor_size = cls.dim_to_size(tensor_to_check.get_dims())
+        x_pos_jacobian = np.zeros((tensor_size, 0), dtype=np.float64)
+        x_neg_jacobian = np.zeros((tensor_size, 0), dtype=np.float64)
+
+        def concat_flatten_output(row, jacobian_matrix):
+            if jacobian_matrix.shape[1] == 0:
+                # first time, concate output dynamically
+                output_vals = []
+                for output_name in output_names:
+                    op.run(scope, ctx)
+                    output_val = np.array(
+                        cls.get_tensor_by_name(scope, output_name)).flatten()
+                    output_vals = np.append(output_vals, output_val)
+                # get dimension info, allocate memory
+                jacobian_matrix.resize(
+                    (tensor_size, len(output_vals)), refcheck=False)
+                jacobian_matrix[row, :] = output_vals.flatten()
+            else:
+                start_idx = 0
+                for output_name in output_names:
+                    op.run(scope, ctx)
+                    output_val = np.array(
+                        cls.get_tensor_by_name(scope, output_name)).flatten()
+                    jacobian_matrix[row, start_idx:start_idx+len(output_val)] \
+                            = output_val
+                    start_idx += len(output_val)
+
+        for i in xrange(tensor_size):
+            if in_place:
+                cls.feed_input(scope, op, inputs, cpu_place)
+            origin_val = tensor_to_check.get_float_element(i)
+            x_pos = origin_val + delta
+            tensor_to_check.set_float_element(i, x_pos)
+            concat_flatten_output(i, x_pos_jacobian)
+            if in_place:
```

Comment:
I'm not sure about this line. Review of someone else is needed.

Reply:
Me too, just to keep consistent with original logic.